### ISSUE TITLE
three arguments with same type is confusing

### ISSUE THREAD

**Issue by hax** [Richienb/proposal-math-clamp#4]
The function currently includes three arguments of the same type, making it difficult to determine the specific meanings of each parameter when calling the function as `Math.clamp(a, b, c)`. Both the parameter order (value, min, max) and (min, value, max) can confuse some developers because they may expect the opposite.

It might be worth considering alternative approaches such as `Math.clamp(value, { min, max })` or `value.clamp(min, max)`. These alternatives provide a clearer way of expressing the function parameters.

**Comment by Richienb** [comment]
`value.clamp` would obviously be out of the question because there only exist `to*` methods on `Number.prototype`.

As for `{min, max}`, objects are already used in some navigator functions so it wouldn't be that unjavascripty. It would also allow either min or max to be left unspecified so as to allow for more explicit forms: `Math.clamp(value, {min: 0}) === Math.max(0, value)`.

**Comment by Richienb** [comment]
Following https://github.com/Richienb/proposal-math-clamp/issues/5#issuecomment-1895781703, we can name the arguments `number`, `min` and `max`.

**Comment by Richienb** [comment]
What about if neither `min` nor `max` is provided by the user?
What about if no object at all is provided by the user?

Should `Math.clamp` return the number untouched, or throw a TypeError?

**Comment by Richienb** [comment]
> (I also don't find the use-case of omitting "min" or "max" arguments overly compelling either, since we already have a `Math.min()` and `Math.max()` function to handle that use-case. Perhaps it might be useful to dynamically omit the min or max side based on some condition, but that's already possible with `Infinity`, e.g. `Math.clamp(value, 0, condition ? 360 : Infinity)`.)

The use case is to decrease cognitive load.

Before:

```js
// Clamp score to a lower bound of 0
Math.max(score, 0);
```

After:

```js
math.clamp(score, {min: 0});
```

**Comment by hax** [comment]
> you'd still be able to easily figure out what the order is by reading the code.

While carefully reading the code can help understand the purpose of each parameter, I still believe that having all parameters as variables can be problematic during code reviews, especially when there is a mismatch between the intention and the actual invocation (i.e., how can you tell if a code order is actually incorrect?).

Furthermore, since the motivation stated in the README is to emulate CSS `clamp`, using a parameter order different from CSS `clamp` could potentially be confusing.

If we completely disregard this point, I speculate that most people would prefer to have the `value` parameter in the first position. However, even with that, I always strive to minimize such cognitive overhead.

**Comment by Richienb** [comment]
Review:

1. Defining `clamp` on the prototype like `number.clamp(min, max)` (a la https://doc.rust-lang.org/std/primitive.f64.html#method.clamp)
Pro: Order is unambiguous
Con: Not possible to only specify max

1. Defining `clamp` on the prototype like `number.clamp({min, max})`
Pro: Order is unambiguous, allows only specifying max
Con: Options object maybe clutter

1. Options object like `Math.clamp(number, {min, max})`
Pro: Order is unambiguous, allows only specifying max
Con: Options object maybe clutter

1. `Math.clamp(number, min, max)`
Con: Possible confusion, not possible to only specify max

1. `Math.clamp(min, number, max)`
Con: Possible confusion, not possible to only specify max

**Comment by hax** [comment]
Using `{min, max}` as an option bag does have a drawback in the context of the `Math.xxx` methods. Although option bags are not rare in the language's standard library, they are not prevalent in the `Math.xxx` methods specifically.

On the other hand, introducing `Number.prototype.clamp` also goes against the convention in some sense since all the existing mathematical functions are under the `Math` object rather than `Number.prototype`.

To me, both approaches break conventions in their own way. However, I personally lean towards `Number.prototype.clamp` because `value.clamp(min, max)` is more concise and simpler to write compared to `Math.clamp(value, { min, max })`.

Additionally, instance methods are easily expandable to `bigInt.clamp()`, whereas it would be a question whether `Math.clamp` would support `bigint` since none of the `Math` methods currently do. Although some TC39 representatives have expressed a desire to extend the `Math` methods to support `bigint`, I anticipate that this is a challenging and potentially impractical problem. Considering future additions like `Decimal`, in the long run, it might become unavoidable to move the computational methods to the prototypes.

**Comment by Richienb** [comment]
@hax I have revised it into a table

It seems that the most unconventional ones are the most useful
Are you familiar with how this sort of trade-off is made?

**Comment by hax** [comment]
@Richienb 

To be honest, each TC39 representatives are always different, making it difficult to determine which trade-off will ultimately be accepted. Sometimes, the outcome relies on the determination of the champion and the strength of opposition. Therefore, based on my observations, the final form of a proposal can sometimes be somewhat arbitrary. 🤪

**Comment by Richienb** [comment]
I just added number.`clamp`/`clampMin`/`clampMax` to the table, and it seems like it covers everything...

**Comment by Richienb** [comment]
Virtually all userland implementations order them like (number, min, max) which makes me think that in practice, order ambiguity might not be as big of a problem if we just order them like that as well.

**Comment by theScottyJam** [comment]
I'm going to retract this statement I said earlier:

> (I also don't find the use-case of omitting "min" or "max" arguments overly compelling either, since we already have a Math.min() and Math.max() function to handle that use-case. Perhaps it might be useful to dynamically omit the min or max side based on some condition, but that's already possible with Infinity, e.g. Math.clamp(value, 0, condition ? 360 : Infinity).)

I've gotten pretty used to reading and writing code that uses `Math.min()` and `Math.max()` to handle clamping, and I don't have any practice with a clamp() method yet, so it took me a while to come around to it, but I can see now how it's easier to read understand code that's uses an explicit `clamp()` function with some arguments omitted as opposed to using min() or max() for clamping. With a clamp function, regardless of how the final function signature looks like, it's pretty easy to read it and picture "this parameter is the lower bound while this one is the upper bound", while with something like `Math.max(n, 10)` it does take me a moment to figure out if the `10` is acting like a lower or upper bound - I basically have to run an example scenario through my head to remember how it works (I'm taking the max of these two numbers, so if n is, say, 20, then the max of 20 and 10 would be 20, which means the 10 would be operating as a lower bound).

**Comment by michaelficarra** [comment]
I claim that CSS parameter order (lower bound, number to clamp, upper bound) is the only *generally intuitive* order. What I mean by that is the order that someone would guess the parameters to be in with no other context or programming experience. To my knowledge, across cultures, even those using right-to-left writing systems, real numbers are depicted on a number line going from low numbers on the left to high numbers on the right. Using the number line as intuition, it makes sense that you would pass clamp a low bound on the left and a high bound on the right, with the number that is supposed to be between them... between them.

Failing agreement on that, I think we'd have to end up going with the prototype method, which again, *obviously* would take the lower bound on the left and the upper bound on the right.

**Comment by Richienb** [comment]
@michaelficarra I noticed that despite your "intuitive" order being a part of the spec for some time, every single commenter used (num, min, max) order anyways. Not a single person assumed the "intuitive" order was the actual order, even though it was. So, I changed it to the one they were using.

**Comment by theScottyJam** [comment]
The number-line example does fall apart when you consider the fact that the value you're trying to clamp isn't actually clamped yet. It could be smaller than the minimum number or larger than the maximum number. It'll only be between the two numbers after the clamping operation has finished.

There are other intuitive reasons one would expect the value being clamped to come first. And I think it's fair to use a programmer's intution here since we are programmers.

When we write static methods, we typically place the subject of the method as the first argument. Some examples:
* `Object.hasOwn(user, 'name')`
* `Object.assign(user, userChanges)`
* `Array.from(iterator, mapFn)`

There's a really strong presedence for doing this sort of thing. Clamping isn't all that different in this regard - the value we're trying to clamp would come first, and then other parameters explaining how the clamping operation should be performed would come afterwards.

That being said, I see what you're saying and understand why you feel it is intuitive. But there's other "intuitive" ways to look at this issue.

**Comment by bakkot** [comment]
The natural-language interpretation of this function is "clamp `x` between `y` and `z`". That is, we're operating on `x`. That is much much more relevant than thinking about a number line. That is why every single programming language with a three-argument `clamp` uses the (num, min, max) form. We are not going to break precedent with every other language.

**Comment by Richienb** [comment]
Updated comparison, after ruling out option bags and multiple functions from https://github.com/CanadaHonk/proposal-math-clamp/issues/4#issuecomment-1897667310

|Method|Unambiguous order|Easily extendable to BigInt|Conventional|
|-|:-:|:-:|:-:|
|`number.clamp(min, max)`|✔️|✔️|🟡 Number prototype, diverges from `Math.min`/`max`|
|`Math.clamp(number, min, max)`|❌|✔️|✔️|
|`Math.clamp(min, number, max)`|❌|✔️|✔️✔️ CSS signature|

**Comment by Richienb** [comment]
I believe the argument for `(number, min, max)` vs `(min, number, max)` is: how much would we be willing to trade off consistency within the language with conventions across most other languages?

**Comment by bakkot** [comment]
One of the people who originally added the CSS function has said that [with hindsight the weight of precedent suggested other order](https://matrixlogs.bakkot.com/TC39_Delegates/2025-02-18#L232), [CSS functions aren't much like JS functions anyway](https://matrixlogs.bakkot.com/TC39_Delegates/2025-02-18#L247), and [we should just do the obvious thing in JS](https://matrixlogs.bakkot.com/TC39_Delegates/2025-02-18#L246) i.e. [`(val, min, max)`](https://matrixlogs.bakkot.com/TC39_Delegates/2025-02-18#L266).

We should not be willing to break precedent with every other language just to match CSS here.

**Comment by woody-li** [comment]
Vote +1 for `Number.prototype.clamp`

**Comment by js-choi** [comment]
For what it’s worth, my [old Stage 0 proposal-math-clamp](https://github.com/js-choi/proposal-math-clamp) was firmly on the side of the `Math.clamp(x, minX, maxX)` design:

> The precedent in C++, C#, Lua, and OpenGL all seems to have the interval limits always be inclusive on both sides.

(I don’t think that the relationship between CSS API design and core JavaScript language API design is super strong, especially if CSS’s design is quirky or a “mistake”. As an example, CSS’s index ordinals start from 1, not 0…)

**Comment by bbb651** [comment]
Pico8 has a really elegant solution to this, it has a [`mid(first, second, third)`](https://breezewiki.catsarch.com/pico-8/wiki/Mid) function that just returns the middle number. If you think of any two arguments as `min` and `max`, the 3rd one is either in between them and will be the middle value, `<= min` which makes `min` the middle value or `>= max` which makes `max` the middle value.

Although I think I'm still in favor of `Number.prototype.clamp`, having both `Math.clamp(min, val, max)` and `Math.clamp(val, min, max)` (and others) possible makes it code style dependent and harder to infer from usage.

**Comment by gameroman** [comment]
> Pico8 has a really elegant solution to this, it has a [`mid(first, second, third)`](https://breezewiki.catsarch.com/pico-8/wiki/Mid) function that just returns the middle number. If you think of any two arguments as `min` and `max`, the 3rd one is either in between them and will be the middle value, `<= min` which makes `min` the middle value or `>= max` which makes `max` the middle value.
> 
> Although I think I'm still in favor of `Number.prototype.clamp`, having both `Math.clamp(min, val, max)` and `Math.clamp(val, min, max)` (and others) possible makes it code style dependent and harder to infer from usage.

I actually like like this a lot, manual `clamp` implementations are already "code style dependent", and this makes it simpler to not have to think about which order you have to pass the arguments in and also solves the problem of not having to do `Number.prototype.clamp`

**Comment by gameroman** [comment]
So something like this
```ts
function clamp(a: number, b: number, c: number): number {
  return a > b ? (b > c ? b : a > c ? c : a) : a > c ? a : b > c ? c : b;
}
```

Though I also like `Number.prototype.clamp` too

**Comment by ljharb** [comment]
imo making argument order not matter is horrifically bad API design (`.substring` does this, and i hate it).

**Comment by michaelficarra** [comment]
@ljharb I don't really think you can make such a general claim. When the arguments are *semantically distinct*, sure. But nobody would say that `Math.min` and `Math.max` accepting their arguments in any order is bad API design. This proposed operation takes 3 semantically-identical numbers in any order, just like `+` and `*` take 2 semantically-identical numbers in any order.

**Comment by ljharb** [comment]
That's a fair qualifier, but clamp doesn't take 3 semantically identical numbers. The upper bound, lower bound, and input are three distinct items.

**Comment by michaelficarra** [comment]
@ljharb By "this proposed operation", I was referring to the `mid` function that has been discussed in the recent comments on this thread (which @gameroman confusingly named `clamp` in their most recent code example).

@gameroman @bbb651 I like the ergonomics of `mid` a lot, but the one downside of it is that it will make 8/3 comparisons in the average call whereas `clamp` will be guaranteed to make only 2 comparisons. But after accounting for the additional handling of NaNs and -0, they're basically the same. And as a bonus, it should cover additional use cases that a simple clamping function would not. So I would support this proposal pivoting to a `mid` function.

**Comment by bakkot** [comment]
I appreciate the cleverness of `mid` but I'm concerned it's not going to be clear to readers. Like, if someone reads `return array[Math.mid(index, 0, array.length - 1)]`, I expect the immediate reaction is "what is going on", whereas if it were `Math.clamp` it would be completely obvious.

**Comment by ljharb** [comment]
"mid" sounds like a very confusing operation, and a different proposal that I don't think has sufficient motivation to be added to the language. This proposal needs "clamp".

**Comment by michaelficarra** [comment]
> I appreciate the cleverness of `mid` but I'm concerned it's not going to be clear to readers.

Yeah, I think working on naming could help with that. `median3`?. If we wanted to, we could even call it `clamp` since it entirely subsumes that interface.

> a different proposal that I don't think has sufficient motivation

I don't understand this feedback. A `mid` function provides exactly the interface that many in this issue tracker have advocated for for `Math.clamp` (regardless of their preferred parameter order). It has the side benefit of *also* coving other use cases that clamp alone would not, such as the median-of-three pivot selection in Quicksort.

**Comment by ljharb** [comment]
I don't think it's a benefit if those other use cases don't have a sufficient driving userland need. "Clamp" is what needs solving and what has stage 1, if "median-of-three" needs solving, a different proposal is welcome to attempt it.

**Comment by theScottyJam** [comment]
[Bakkot's comment](https://github.com/tc39/proposal-math-clamp/issues/4#issuecomment-4917552565) convinces me that a mid function would hurt readability, and as such should probably not be the route we take.

But I feel like ideas like these shouldn't be hand-waved away as out-of-scope simply because it does more than required to solve the problem. Sometimes the best way to solve a specific problem is by solving a more general problem (I.e. [the inventor's paradox](https://en.wikipedia.org/wiki/Inventor%27s_paradox)), and there's nothing wrong with going back to the committee and saying that we'd like to expand on what the proposal does, because we believe a more general API is better. It seems to happen fairly often - if memory serves, earlier versions of groupBy() were trying to simply solve a partitioning use case, but they pivoted to the more general groupBy() API when they realized the more general solution solved partitioning very well, and could do more.

**Comment by gameroman** [comment]
Yes, what I'm saying is that `Math.clamp` would still be called `Math.clamp` but could behave like `mid`

**Comment by hax** [comment]
I quite like the idea of `mid`. It not only solves this issue (you can use arbitrary parameter order with `mid`), but also addresses the problem of swapped boundary values (whether to take `max`, `min`, or throw an exception), avoiding mental burden (for both humans and coding agents). Perhaps the only concern is that `mid` might be less common than `clamp`—when I need to constrain a value within upper and lower bounds, I might not immediately think of `mid` as the right solution. However, given that there is currently no standard solution in the JavaScript ecosystem either, this learning cost may not pose a real obstacle.

**Comment by theScottyJam** [comment]
Another slightly problematic point to a `.mid()` (or a `.clamp()` that acts like `.mid()`) is that parameter order becomes a convention, and different people can follow different conventions. For example:

```javascript
character.pos.x = Math.mid(character.pos.x, 0, SCREEN_WIDTH);
// vs
character.pos.x = Math.mid(0, character.pos.x, SCREEN_WIDTH);
```

Perhaps both lines of code technically act the same, but regardless of how we order them, conceptually, the mental model is that we are clamping `character.pos.x` between `0` and `SCREEN_WIDTH`. The mental model isn't "use the value in the middle".
* This makes reading the code slightly more difficult - every time you read `Math.mid()`, there's a bit of extra burden in trying to figure out which parameters are intended to be the bounds, and which one is the value being clamped, information that's needed as you build your mental model for this line of code. The bound parameters are somewhat obvious in the above example, but that might not always be as clear in real-world code, and it might require a few seconds to figure it out.
* If you want to enforce a particular parameter order in a given codebase, good luck. It's not something a linter can easily help out with.

**Comment by SmashingQuasar** [comment]
Could we not have Javascript reinvent a concept that is already massively implemented and supported in most popular languages and simply abide to the overall consensus on this, please?
At minimum C++ and C# are using `clamp(value, min, max)`. It is not difficult. If this is really outside of the realm of comprehension by Javascript users, then `value.clamp(min, max)` but it will be contradictory to how `Math` functions.
Let's not create another `.substr()` and `.substring()` disaster.
